bkp/bkp_ekat_assert.hpp

```diff
@@ -0,0 +1,146 @@
+#ifndef EKAT_ASSERT_HPP
+#define EKAT_ASSERT_HPP
+
+#include <sstream>
+#include <exception>
+#include <assert.h>
+#include <stdexcept>  // For std::logic_error
+
+#include <cfenv>  // For FPE environment
+
+/*
+ * Asserts and error checking for Scream.
+ *
+ * scream_kernel_* are for error checking within kokkos kernels.
+ *
+ * Any check with "assert" in the name is disabled for release builds
+ *
+ * For _msg checks, the msg argument can contain '<<' if not a kernel check.
+ */
+
+// Internal do not call directly
+#define EKAT_THROW(msg, exception_type)                  \
+    std::stringstream _ss_;                              \
+    _ss_ << __FILE__ << ":" << __LINE__ << ": FAIL:\n";  \
+    _ss_ << "\n" << msg;                                 \
+    throw exception_type(_ss_.str());                    \
+
+#define EKAT_CHECK(condition, false_reaction)     \
+  do {                                            \
+    if ( ! (condition) ) {                        \
+      impl_throw(msg #condition, exception_type); \
+    }                                             \
+  } while(0)
+
+
+#define EKAT_KERNEL_THROW(condition, msg)             \
+  do {                                          \
+    if ( ! (condition) )                        \
+      Kokkos::abort(#condition "\n" msg);       \
+  } while (0)
+
+#ifndef NDEBUG
+
+#define EKAT_ASSERT(condition)                 EKAT_CHECK(condition, "",  std::logic_error)
+#define EKAT_ASSERT_MSG(condition, msg)        EKAT_CHECK(condition, msg, std::logic_error)
+
+#define EKAT_KERNEL_ASSERT(condition)          EKAT_KERNEL_CHECK(condition, "")
+#define EKAT_KERNEL_ASSERT_MSG(condition, msg) EKAT_KERNEL_CHECK(condition, msg)
+
+#else
+
+#define ekat_assert(condition)  ((void) (0))
+#define ekat_assert_msg(condition, msg)  ((void) (0))
+
+#define ekat_kernel_assert(condition) ((void) (0))
+#define ekat_kernel_assert_msg(condition, msg) ((void) (0))
+
+#endif
+
+#define ekat_require(condition)                 impl_throw(condition, "",  std::logic_error)
+#define ekat_require_msg(condition, msg)        impl_throw(condition, msg, std::logic_error)
+
+#define ekat_kernel_require(condition)          impl_kernel_throw(condition, "")
+#define ekat_kernel_require_msg(condition, msg) impl_kernel_throw(condition, msg)
+
+#define ekat_error_msg(msg)                     scream_require_msg(false, msg)
```

Comment:
This shouldn't dispatch to scream_require_msg, should it?

Reply:
why isn't my git grep showing these? Thanks!

Reply:
You're just trying to make me feel valued as a review!